external-crates/move/crates/move-vm-runtime/src/execution/interpreter/locals.rs

```diff
@@ -0,0 +1,305 @@
+                    .with_message(format!("Invalid index: {}", ndx))
+            })
+    }
+}
+
+// -------------------------------------------------------------------------------------------------
+// Machine (Runtime) Heap
+// -------------------------------------------------------------------------------------------------
+
+impl Default for MachineHeap {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl MachineHeap {
+    pub fn new() -> Self {
+        Self {
+        }
+    }
+
+    /// Allocates a stack frame with the given size.
+    /// If there is not enough space in the heap, it returns an error.
+    pub fn allocate_stack_frame(
+        &mut self,
+        params: Vec<Value>,
+        size: usize,
+    ) -> PartialVMResult<StackFrame> {
+        // Calculate how many invalid values need to be added
+        let invalids_len = size - params.len();
+
+        // Initialize the stack frame with the provided parameters and fill remaining slots with `Invalid`
+        let local_values = params
+            .into_iter()
+            .chain((0..invalids_len).map(|_| Value::invalid())) // Fill the rest with `Invalid`
+            .collect::<Vec<Value>>();
+
+        Ok(StackFrame { slice: local_values })
+    }
+
+    /// Frees the given stack frame, ensuring that it is the last frame on the heap.
+    pub fn free_stack_frame(&mut self, _frame: StackFrame) -> PartialVMResult<()> {
+        Ok(())
+    }
+}
+
+// -------------------------------------------------------------------------------------------------
+// Stack Frame
+// -------------------------------------------------------------------------------------------------
+
+impl StackFrame {
+    pub fn iter(&self) -> std::slice::Iter<'_, Value> {
+        self.slice.iter()
+    }
+
+    /// Makes a copy of the value, via `value.copy_value`
+    pub fn copy_loc(&self, ndx: usize) -> PartialVMResult<Value> {
+        if self.is_invalid(ndx)? {
+            return Err(
+                PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                    .with_message("Cannot copy from an invalid memory location".to_string()),
+            );
+        }
+        self.slice
+            .get(ndx)
+            .ok_or_else(|| {
+                PartialVMError::new(StatusCode::INTERNAL_TYPE_ERROR)
+                    .with_message(format!("Local index out of bounds: {}", ndx))
+            })
+            .map(|value| value.copy_value())
+    }
+
+    /// Moves a location out of memory, swapping it with `ValueImpl::Invalid`
+    pub fn move_loc(&mut self, ndx: usize) -> PartialVMResult<Value> {
+        if self.is_invalid(ndx)? {
+            return Err(
+                PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                    .with_message("Cannot move from an invalid memory location".to_string()),
+            );
+        }
+
+        let value = std::mem::replace(
```

Comment:
Wondering if there should be a `move_value` or `take_value` function on `Value` that would do this, just from an API perspective on the `Value` type.

Reply:
Values do not really exist that way. Values can only be moved or taken if they have a location to be moved or taken from, and so we model that on the locations themselves. Taking or moving a value does not make much sense, as a value ref, e.g., may live in a location and not know it.